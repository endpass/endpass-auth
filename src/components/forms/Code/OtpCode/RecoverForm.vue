<template>
  <form
    data-test="recover-form"
    @submit.prevent="onSubmit"
  >
    <v-title>
      <span v-html="$t('components.recover.enterSeed')" />
    </v-title>
    <form-item class="v-mb-24">
      <v-input
        v-model="seedPhrase"
        :error="error"
        label="Seed phrase"
        name="seedPhrase"
        :placeholder="$t('components.recover.seedPhrase')"
        required
        data-test="seed-phrase"
      />
    </form-item>
    <form-row>
      <v-button
        type="submit"
        :disabled="!isSeedPhraseValid || isLoading"
        :is-loading="isLoading"
        data-test="submit-button"
      >
        {{ $t('global.confirm') }}
      </v-button>
    </form-row>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import VInput from '@endpass/ui/kit/VInput';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import { authStore, coreStore } from '@/store';
import VTitle from '@/components/common/VTitle';

export default {
  name: 'RecoverForm',

  authStore,
  coreStore,

  data: () => ({
    error: null,
    seedPhrase: '',
    isLoading: false,
  }),

  computed: {
    isSeedPhraseValid() {
      return this.seedPhrase.length >= 12;
    },
  },

  methods: {
    async onSubmit() {
      if (this.isLoading) return;
      try {
        this.isLoading = true;
        await this.$options.authStore.disableOtp({
          seedPhrase: this.seedPhrase,
        });
        this.$emit('submit');
      } catch (err) {
        console.error(err);

        const msg =
          (err && err.message) ||
          this.$i18n.t('components.otpBlock.recoverFailed');
        this.error = msg;
      } finally {
        this.isLoading = false;
      }
    },
  },

  components: {
    VTitle,
    VButton,
    VInput,
    FormItem,
    FormRow,
  },
};
</script>
