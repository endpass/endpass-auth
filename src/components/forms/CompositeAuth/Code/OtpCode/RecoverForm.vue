<template>
  <form
    data-test="recover-form"
    @submit.prevent="onSubmit"
  >
    <v-title
      data-test="form-title"
      :html="$t('components.recover.enterSeed')"
    />
    <form-item is-last>
      <v-input
        v-model="seedPhrase"
        :error="error"
        label="Seed phrase"
        name="seedPhrase"
        :placeholder="$t('components.recover.seedPhrase')"
        required
        data-test="seed-phrase"
      />
    </form-item>
    <form-row>
      <v-button
        type="submit"
        :disabled="!isSeedPhraseValid || isLoading"
        data-test="submit-button"
      >
        {{ primaryButtonLabel }}
      </v-button>
    </form-row>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import VInput from '@endpass/ui/kit/VInput';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import { authStore, coreStore } from '@/store';
import VTitle from '@/components/common/VTitle';

export default {
  name: 'RecoverForm',

  authStore,
  coreStore,

  data: () => ({
    error: null,
    seedPhrase: '',
  }),

  computed: {
    isLoading() {
      return this.$options.coreStore.loading;
    },

    primaryButtonLabel() {
      return !this.isLoading
        ? this.$i18n.t('global.confirm')
        : this.$i18n.t('global.loading');
    },

    isSeedPhraseValid() {
      return this.seedPhrase.length >= 12;
    },
  },

  methods: {
    async onSubmit(seedPhrase) {
      try {
        await this.$options.authStore.disableOtp({ seedPhrase });
        this.$emit('recover');
      } catch (err) {
        console.error(err);

        const msg =
          (err && err.message) ||
          this.$i18n.t('components.otpBlock.recoverFailed');
        this.error = msg;
      }
    },
  },

  components: {
    VTitle,
    VButton,
    VInput,
    FormItem,
    FormRow,
  },
};
</script>
