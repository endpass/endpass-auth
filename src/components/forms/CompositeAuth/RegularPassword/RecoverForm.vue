<template>
  <form @submit.prevent="onSubmit">
    <v-title>
      <span v-html="$t('components.regularPasswordRecover.title')" />
    </v-title>
    <form-item>
      <v-input
        v-model="password"
        v-validate="'required|min:8'"
        type="password"
        required
        data-vv-as="password"
        data-vv-name="password"
        :error="errors.first('password')"
        name="password"
        :placeholder="$t('components.regularPasswordRecover.newPassword')"
        data-test="password-input"
      />
    </form-item>
    <form-item>
      <v-input
        v-model="repeatPassword"
        v-validate="'required|min:8'"
        type="password"
        required
        data-vv-as="password"
        data-vv-name="repeatPassword"
        :error="errors.first('repeatPassword')"
        name="repeatPassword"
        :placeholder="$t('components.regularPasswordRecover.repeatPassword')"
        data-test="repeat-password-input"
      />
    </form-item>
    <form-item>
      <v-input
        v-model="code"
        v-validate="'required|digits:6'"
        data-vv-as="code"
        data-vv-name="code"
        :error="errors.first('code')"
        name="code"
        :label="$t('components.regularPasswordRecover.labelCode')"
        :placeholder="$t('components.regularPasswordRecover.placeholderCode')"
        data-test="password-input"
      />
    </form-item>
    <form-item class="v-mb-24">
      <form-controls>
        <v-button
          :disabled="isLoading"
          skin="quaternary"
          data-test="submit-button"
          @click="closeForm"
        >
          {{ $t('global.cancel') }}
        </v-button>
        <v-button
          :disabled="!isSubmitEnable || isLoading"
          :is-loading="isLoading"
          type="submit"
          data-test="submit-button"
        >
          {{ $t('global.confirm') }}
        </v-button>
      </form-controls>
    </form-item>
    <form-row class="v-fs-14 v-text-center">
      {{ $t('components.regularPasswordRecover.didntGetCode') }}&nbsp;
      <v-link
        :disabled="isLoading"
        href="#"
        data-test="send-code"
        @click.prevent="sendCode"
      >
        {{ $t('components.regularPasswordRecover.sendTitle') }}
      </v-link>
    </form-row>
  </form>
</template>

<script>
import VButton from '@endpass/ui/kit/VButton';
import VInput from '@endpass/ui/kit/VInput';
import VLink from '@endpass/ui/kit/VLink';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import formMixin from '@/mixins/form';
import { authStore } from '@/store';
import VTitle from '@/components/common/VTitle';
import FormControls from '@/components/common/FormControls';

export default {
  name: 'PasswordForm',

  authStore,

  props: {
    email: {
      type: String,
      required: true,
    },
  },

  data: () => ({
    password: '',
    repeatPassword: '',
    code: '',
    isLoading: false,
  }),

  computed: {
    isSubmitEnable() {
      return this.isFormValid && this.isPasswordEqual;
    },

    isPasswordEqual() {
      return this.password && this.password === this.repeatPassword;
    },
  },

  methods: {
    async onSubmit() {
      if (this.isLoading) return;
      try {
        this.isLoading = true;
        this.$validator.errors.removeById('sendCodeId');
        await this.$options.authStore.confirmResetRegularPassword({
          password: this.password,
          code: this.code,
        });
        this.$emit('submit', this.password);
      } catch (error) {
        this.$validator.errors.add({
          field: 'code',
          msg: this.$i18n.t('components.regularPasswordRecover.recoveryError'),
          id: 'sendCodeId',
        });
      } finally {
        this.isLoading = false;
      }
    },

    closeForm() {
      this.$emit('close');
    },

    async sendCode() {
      if (this.isLoading) return;
      try {
        this.$validator.errors.removeById('sendCodeId');
        this.isLoading = true;
        await this.$options.authStore.sendCode({ email: this.email });
      } catch (error) {
        this.$validator.errors.add({
          field: 'code',
          msg: this.$i18n.t('components.regularPasswordRecover.sendError'),
          id: 'sendCodeId',
        });
      } finally {
        this.isLoading = false;
      }
    },
  },

  mounted() {
    this.sendCode();
  },

  mixins: [formMixin],

  components: {
    FormControls,
    VLink,
    VTitle,
    VButton,
    VInput,
    FormItem,
    FormRow,
  },
};
</script>

<style lang="postcss"></style>
