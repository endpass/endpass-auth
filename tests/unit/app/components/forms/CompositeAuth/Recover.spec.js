import { shallowMount, createLocalVue } from '@vue/test-utils';
import RecoverForm from '@/components/forms/CompositeAuth/Recover.vue';

import setupI18n from '@/locales/i18nSetup';

const localVue = createLocalVue();
const i18n = setupI18n(localVue);
describe('Recover', () => {
  const seedPhrase = 'foo bar foo bar foo bar foo bar foo bar foo bar';
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(RecoverForm, {
      provide: {
        theme: 'default',
      },
      localVue,
      i18n,
    });
  });

  describe('render', () => {
    it('should correctly render component', () => {
      expect(wrapper.name()).toBe('RecoverForm');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render error', () => {
      const error = 'error';

      wrapper.setProps({ error });

      expect(wrapper.find('v-input-stub').attributes('error')).toBe(error);
    });

    it('should correctly disable submit button', () => {
      const submitButton = wrapper.find('[data-test=submit-button]');

      wrapper.setProps({
        loading: true,
      });

      expect(submitButton.text()).toBe('Loading...');
      expect(submitButton.attributes().disabled).toBeTruthy();

      wrapper.setProps({
        loading: false,
      });

      expect(submitButton.text()).toBe('Confirm');
      expect(submitButton.attributes().disabled).toBeTruthy();

      wrapper.setProps({
        loading: false,
      });
      wrapper.setData({ seedPhrase });

      expect(submitButton.text()).toBe('Confirm');
      expect(submitButton.attributes().disabled).toBeUndefined();
    });
  });

  describe('behavior', () => {
    it('should not allow submit form if seed phrase is invalid', () => {
      const form = wrapper.find('form');

      wrapper.setData({
        seedPhrase: '',
      });

      form.trigger('submit');
      expect(wrapper.emitted().submit).toBe(undefined);

      wrapper.setData({
        seedPhrase: 'foo bar',
      });

      form.trigger('submit');
      expect(wrapper.emitted().submit).toBe(undefined);
    });

    it('should submit form', () => {
      wrapper.setData({ seedPhrase });

      wrapper.find('form').trigger('submit');
      expect(wrapper.emitted().submit).toEqual([[seedPhrase]]);
    });
  });
});
