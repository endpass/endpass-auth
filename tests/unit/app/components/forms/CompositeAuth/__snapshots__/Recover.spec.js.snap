// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recover render should correctly render component 1`] = `
<form>
  <form-field-stub>
    <message-stub data-test="form-message" class="v-modal-card-title">
      Enter recovery seed phrase of your wallet
    </message-stub>
  </form-field-stub>
  <form-field-stub>
    <v-input-stub label="Seed phrase" value="" name="seedPhrase" placeholder="Seed phrase" required=""></v-input-stub>
  </form-field-stub>
  <form-controls-stub>
    <v-button-stub skin="primary" size="normal" type="submit" disabled="true" data-test="submit-button">
      Confirm
    </v-button-stub>
  </form-controls-stub>
</form>
`;
